docs: update changelog guide to use CREATE SOURCE TABLE (#8458)
confluentinc/ksql: docs/how-to-guides/convert-changelog-to-table.md

## Materializing a changelog topic

If you have a `changelog` topic, and you want a view of the data that reflects the latest values for each key then simply create a table with the `changelog` topic using the `CREATE SOURCE TABLE` statement.

Let's say that you have the following data in your `changelog` topic where the first row is the record that has the earliest offset and the last row is the record that has the latest offset:
```sql
+------+------+------+------+
|K     |V1    |V2    |V3    |
+------+------+------+------+
|k1    |0     |a     |true  |
|k2    |1     |b     |false |
|k1    |2     |c     |false |
|k3    |3     |d     |true  |
|k2    |4     |e     |true  |
```
Note that there are 2 records for the keys `k1` and `k2` in the `changelog` topic.

Make a table `latest_view` with four columns. `k` represents the key of the table. Rows with the same key represent information about the same entity. `v1`, `v2`, and `v3` are various value columns.
```sql

> CREATE SOURCE TABLE latest_view (

Reviewer: @spena I cannot remember but does `CST` respect `auto.offset.reset` config? If yet, we should add that it should be set to `"earliest"` -- if it ignores the config and always reads from earliest, we should also mention it IMHO.

Reviewer 2: It respects that config. The initial idea was to read always from earliest, but we decided to respect it instead.

    k VARCHAR PRIMARY KEY,
    v1 INT,
    v2 VARCHAR,
    v3 BOOLEAN
) WITH (
    kafka_topic = 'changelog',
    partitions = 1,
    value_format = 'JSON'
);
```

> Now, you can view the latest values for each key in your `changelog` topic by issuing a pull query against the `latest_view` table that we just created above:

Reviewer: Might be good to mention, that if there is a lot of data in the topic, it might take a while until the table contains the latest data -- pull query are accepted right way but might return "old" data as long as the end of the input topic is not reached.

> Now, you can view the latest values for each key in your `changelog` topic by issuing a pull query against the `latest_view` table that we just created above:

Reviewer 3: ```suggestion
Now, you can view the latest values for each key in your `changelog` topic by issuing a pull query against the `latest_view` table that you created above:
```

```sql
SELECT * FROM latest_view;
+------+------+------+------+
|K     |V1    |V2    |V3    |
+------+------+------+------+
|k1    |2     |c     |false |
|k2    |4     |e     |true  |
|k3    |3     |d     |true  |
```

Notice how for each key, the columns reflect the latest set of values.

> If you just want to look up the latest value for a particular key (let's assume `k2` in this case), then you can simply issue a pull query for that particular key:

Reviewer 3: ```suggestion
If you want to look up the latest value for a particular key, for example, `k2`, issue a pull query for that key:
```

```sql
SELECT * FROM latest WHERE k = 'k2';
+------+------+------+------+
|K     |V1    |V2    |V3    |
+------+------+------+------+
|k2    |4     |e     |true  |
```

## Materializing a changelog `STREAM`

> The best way to materialize an input topic as a table is the `CREATE SOURCE TABLE` statement. If you already have a `STREAM`, though, you can also materialize it as a table with the `CREATE TABLE ... AS ...` statement. Let's say you have a `STREAM` of events in ksqlDB that represent a series of changes called `changelog_stream`, and you want a view of the data that reflects the latest values for each key. 

Reviewer 3: ```suggestion
The best way to materialize an input topic as a table is the `CREATE SOURCE TABLE` statement. If you already have a `STREAM`, you can also materialize it as a table with the `CREATE TABLE ... AS ...` statement. This is useful if you have a `STREAM` of events in ksqlDB that represents a series of changes called `changelog_stream`, and you want a view of the data that reflects the latest values for each key. 
```

Begin by telling ksqlDB to start all queries from the earliest point in each topic:
```sql
SET 'auto.offset.reset' = 'earliest';
```

Let's mimic adding a few records to the `changelog_stream` by using the `INSERT INTO` statement:
```sql
INSERT INTO changelog_stream (
    k, v1, v2, v3
) VALUES (
    'k1', 0, 'a', true
);

INSERT INTO changelog_stream (
    k, v1, v2, v3
) VALUES (
    'k2', 1, 'b', false
);

INSERT INTO changelog_stream (
    k, v1, v2, v3
) VALUES (
    'k1', 2, 'c', false
);

INSERT INTO changelog_stream (
    k, v1, v2, v3
) VALUES (
    'k3', 3, 'd', true
);

INSERT INTO changelog_stream (
    k, v1, v2, v3
) VALUES (
    'k2', 4, 'e', true
);
```

You can view the data in the `changelog_stream` by issuing a pull query:
```sql
SELECT * FROM changelog_stream;
+------+------+------+------+
|K     |V1    |V2    |V3    |
+------+------+------+------+
|k1    |0     |a     |true  |
|k2    |1     |b     |false |
|k1    |2     |c     |false |
|k3    |3     |d     |true  |
|k2    |4     |e     |true  |

> Query terminated

Reviewer: Do we really print `Query terminated`? Why? For a push query it might make sense, but not for a pull query?

```

To view the data that reflects only the last change for each key in `changelog_stream`, you can derive a table called `latest_events` with a `CREATE TABLE AS SELECT` statement in conjunction with the `LATEST_BY_OFFSET` aggregation:
```sql
CREATE TABLE latest_events AS
    SELECT k,
           LATEST_BY_OFFSET(v1) AS v1,
           LATEST_BY_OFFSET(v2) AS v2,
           LATEST_BY_OFFSET(v3) AS v3
    FROM changelog_stream
    GROUP BY k
    EMIT CHANGES;
```

> Now, you can view the latest values for each key in your `changelog_stream` by issuing a pull query against the `latest_events` table that we just created above:

Reviewer: Maybe add same information that the query will need some time to reach the end of the input stream and that pull queries might return "old" data if the end of the input stream is not reached yet?

> Now, you can view the latest values for each key in your `changelog_stream` by issuing a pull query against the `latest_events` table that we just created above:

Reviewer 3: ```suggestion
Now, you can view the latest values for each key in your `changelog_stream` by issuing a pull query against the `latest_events` table that you created above:
```

```sql
SELECT * FROM latest_events;
+------+------+------+------+
|K     |V1    |V2    |V3    |
+------+------+------+------+
|k1    |2     |c     |false |
|k2    |4     |e     |true  |
|k3    |3     |d     |true  |
```

> Notice how for each key, the columns reflect the latest set of values.

Reviewer: It might also be good to explain, that `LATEST_BY_OFFSET` does not support deletes / tombstones, because it's an aggregation and thus "delete semantics" do not apply.

## Context

You have a topic or a stream of events that represent a series of changes, known as a changelog. You want a view of the data that reflects only the last change for each key. Because ksqlDB represents change over time using tables, you need a way to convert your changelog into a table. This is broadly called *materializing* a changelog into a table.

## Materializing a changelog topic

> If you have a `changelog` topic, and you want view of the data that reflects the latest values for each key then simply create a table with the `changelog` topic using the `CREATE SOURCE TABLE` statement.

Reviewer 4: `and you want view of` missing 'a view'

Let's say that you have the following data in your `changelog` topic where the first row is the record that has the earliest offset and the last row is the record that has the latest offset:
```sql
+------+------+------+------+
|K     |V1    |V2    |V3    |
+------+------+------+------+
|k1    |0     |a     |true  |
|k2    |1     |b     |false |
|k1    |2     |c     |false |
|k3    |3     |d     |true  |
|k2    |4     |e     |true  |
```
Note that there are 2 records for the keys `k1` and `k2` in the `changelog` topic.

Begin by telling ksqlDB to start all queries from the earliest point in each topic:

```sql
SET 'auto.offset.reset' = 'earliest';

> ```

Reviewer 5: This is not necessary with source table, and we shouldn't recommend it here.

Make a table `latest_view` with four columns. `k` represents the key of the table. Rows with the same key represent information about the same entity. `v1`, `v2`, and `v3` are various value columns.
```sql
CREATE SOURCE TABLE latest_view (
    k VARCHAR PRIMARY KEY,
    v1 INT,
    v2 VARCHAR,
    v3 BOOLEAN
) WITH (
    kafka_topic = 'changelog',
    partitions = 1,
    value_format = 'JSON'
);
```

Now, you can view the latest values for each key in your `changelog` topic by issuing a pull query against the `latest_view` table that we just created above:
```sql
SELECT * FROM latest_view;
+------+------+------+------+
|K     |V1    |V2    |V3    |
+------+------+------+------+
|k1    |2     |c     |false |
|k2    |4     |e     |true  |
|k3    |3     |d     |true  |
```

Notice how for each key, the columns reflect the latest set of values.

> If you just want to lookup the latest value for a particular key (let's assume `k2` in this case), then you can simply issue a pull query for that particular key:

Reviewer 5: ```suggestion
If you just want to look up the latest value for a particular key (let's assume `k2` in this case), then you can simply issue a pull query for that particular key:
```

Since there are other changes to make, I'll indulge in a grammar nitpick.

```sql
SELECT * FROM latest WHERE k = 'k2';
+------+------+------+------+
|K     |V1    |V2    |V3    |
+------+------+------+------+
|k2    |4     |e     |true  |
```

## Materializing a changelog `STREAM`

> Let's say you have a `STREAM` of events in ksqlDB that represent a series of changes called `changelog_stream`, and you want view of the data that reflects the latest values for each key. 

Reviewer 4: `you want view of the data` missing `a`

Reviewer 5: ```suggestion
The best way to materialize an input topic as a table is the `CREATE SOURCE TABLE` statement. If you already have a `STREAM`, though, you can also materialize it as a table with the `CREATE TABLE ... AS ...` statement. Let's say you have a `STREAM` of events in ksqlDB that represent a series of changes called `changelog_stream`, and you want a view of the data that reflects the latest values for each key. 
```

Begin by telling ksqlDB to start all queries from the earliest point in each topic:
```sql
SET 'auto.offset.reset' = 'earliest';
```

Let's mimic adding a few records to the `changelog_stream` by using the `INSERT INTO` statement:
```sql
INSERT INTO changelog_stream (
    k, v1, v2, v3
) VALUES (
    'k1', 0, 'a', true
);

INSERT INTO changelog_stream (
    k, v1, v2, v3
) VALUES (
    'k2', 1, 'b', false
);

INSERT INTO changelog_stream (
    k, v1, v2, v3
) VALUES (
    'k1', 2, 'c', false
);

INSERT INTO changelog_stream (
    k, v1, v2, v3
) VALUES (
    'k3', 3, 'd', true
);

INSERT INTO changelog_stream (
    k, v1, v2, v3
) VALUES (
    'k2', 4, 'e', true
);

> ```

Reviewer 5: This setup is missing from the table section. Do you think we need to demonstrate how to create the stream when we don't do the same for the table's input topic?

Author: it is missing because we can't insert into source tables, and I did not want to confuse/distract the readers by adding some other way of inserting rows into the table (eg. with using a helper stream on the topic and inserting into the stream)

Author: it is possible to directly insert into a stream which is why i included it

## Context

> You have a stream of events that represent a series of changes, known as a changelog. You want a view of the data that reflects only the last change for each key. Because ksqlDB represents change over time using tables, you need a way to convert your changelog into a table. This is broadly called *materializing* a changelog into a table.

Reviewer 6: ```suggestion
You have a topic that represent a series of changes, known as a changelog. You want a view of the data that reflects only the last change for each key. Because ksqlDB represents change over time using tables, you need a way to convert your changelog into a table. This is broadly called *materializing* a changelog into a table.
```

Reviewer: > Because ksqlDB represents change over time using tables, you need a way to convert your changelog into a table

This sentence seems miss-leading... if you have a changlog in a topic, you can just call `CREATE [SOURCE] TABLE` to read it.

## In action

```sql

> CREATE TABLE t1 AS

Reviewer: Why not using `CREATE SOURE TABLE` here? Ie, we don't even need the STREAM `changelog`?

    SELECT *

>     FROM changelog

Reviewer: If `changelog` is a STREAM, the updated query is invalid as the result is a `STREAM` and thus you cannot say `CREATE TABLE`.

    EMIT CHANGES;
```

## Materializing a changelog

> In ksqlDB, you derive new tables by aggregating other streams and tables. To create a table that reflects the latest values for each key, declare the `changelog` as a table.

Reviewer 6: this is no longer accurate - you don't need to `aggregate other streams and tables`

Reviewer: Not sure what you mean @agavra? To "convert" a STREAM into a TABLE you can only do it with an aggregation.

Of course, a table-table join also gives you a result table. Thus the sentence is still misleading.

Begin by telling ksqlDB to start all queries from the earliest point in each topic.

```sql
SET 'auto.offset.reset' = 'earliest';
```

Make a table `changelog` with four columns. `k` represents the key of the table. Rows with the same key represent information about the same entity. `v1`, `v2`, and `v3` are various value columns.

```sql

> CREATE TABLE changelog (

Reviewer 5: I wouldn't call the table, "changelog," since it is exactly the inverse of a changelog. Maybe "golegnahc"? Or maybe just "T0", "input_table", etc..

Reviewer: Why do we not use `CREATE SOURE TABLE` ?

    k VARCHAR PRIMARY KEY,
    v1 INT,
    v2 VARCHAR,
    v3 BOOLEAN
) WITH (
    kafka_topic = 'changelog',
    partitions = 1,

>     value_format = 'JSON'

Author: chnaged to json to avoid `Cannot create topic 'changelog' with format AVRO without configuring 'ksql.schema.registry.url'` and remove the need to set up sr

);
```

> Insert some rows into the `changelog` table, repeating entries for some of the keys.

Reviewer: If we switch to `CREATE SOURCE TABLE`, we cannot insert any longer.

Thus, we would need to create a helper STREAM using the same topic and insert into the STREAM.

+-------------------------------------------------------+-------------------------------------------------------+-------------------------------------------------------+-------------------------------------------------------+
|k3                                                     |3                                                      |d                                                      |true                                                   |
```

> You can also run a pull query to scan the entire table to view the last set of values that were inserted for every key.

Reviewer 6: let's just keep this section and delete the individual key part
